### PR TITLE
Introduce flexible scaling

### PR DESCRIPTION
**Pull request by nbalacha**
Allow the cluster to be expanded by one host at a time

Signed-off-by: N Balachandran <nibalach@redhat.com>

### CODE REVIEW

**Comment by nbalacha** [comment]
Starting this off in order to get reviews early. This does not have the TCS changes yet.

**Review by jarrpa — CHANGES_REQUESTED** [review]
Thank you for this! Unfortunately the branch is in a very unclean state.

* The commits still need to be cleaned up.
* You should not be including any changes to the default/managed resources, so please remove those.
* The TSC changes should already be merged, so you probably need a rebase I'm guessing?

**Review by jarrpa — CHANGES_REQUESTED** [review]
Ah, I missed something... don't change the number of StorageClassDeviceSets we generate. That needs to come later after we make a change in Rook-ceph.

File: pkg/controller/storagecluster/cephcluster.go [hunk 1]
```diff
@@ -379,7 +387,7 @@ func newStorageClassDeviceSets(sc *ocsv1.StorageCluster, serverVersion *version.
 			// enable the new behavior while the console is waiting
 			// to be updated.
 			// TODO: Remove this behavior when OCP console is updated
-			count = count / 3
+			count = count / defaults.DeviceSetReplica
```
Comment (jarrpa) [inline pkg/controller/storagecluster/cephcluster.go:390]: This doesn't seem related to this PR, please remove it.

File: pkg/controller/storagecluster/topology.go [hunk 2]
```diff
@@ -264,6 +269,7 @@ func (r *ReconcileStorageCluster) reconcileNodeTopologyMap(sc *ocsv1.StorageClus
 		}
 	}
 
+
```
Comment (jarrpa) [inline pkg/controller/storagecluster/topology.go:272]: Remove this newline.

File: pkg/apis/ocs/v1/storagecluster_types.go [hunk 3]
```diff
@@ -40,6 +40,8 @@ type StorageClusterSpec struct {
 	// ManagedResources specifies how to deal with auxiliary resources reconciled
 	// with the StorageCluster
 	ManagedResources ManagedResourcesSpec `json:"managedResources,omitempty"`
+
+	FlexibleScaling bool            `json:"flexibleScaling,omitempty"`
```
Comment (jarrpa) [inline pkg/apis/ocs/v1/storagecluster_types.go:44]: This needs a comment explaining what the field enables.

**Review by jarrpa — CHANGES_REQUESTED** [review]
Okay, this is much better, and makes sense. Aside from a few small changes, this also needs a test to verify the new functionality works correctly. Add a new test case to `TestFailureDomain()` for flexible scaling.

File: pkg/controller/storagecluster/cephcluster.go (hunk 1 above)
Comment (nbalacha) [reply pkg/controller/storagecluster/cephcluster.go:390]: Ok. I'll send it in a different PR.

File: pkg/controller/storagecluster/topology_test.go [hunk 4]
```diff
@@ -246,7 +271,16 @@ func TestFailureDomain(t *testing.T) {
...
+			label: "Case 3", // storagecluster has predefined failure domain of `host`
+			storageCluster: &api.StorageCluster{
+				Status: api.StorageClusterStatus{
+					FailureDomain: "host",
+				},
+			},
+			expectedFailureDomain: "host",
+		},
+		{
+			label: "Case 4", // storagecluster with three or more zone topology labels
```
Comment (jarrpa) [inline pkg/controller/storagecluster/topology_test.go:283]: Add this as case 5. I see what you're going for, but here I would value the cleaner patch and chronological order over other logical grouping of test cases.

File: pkg/controller/storagecluster/topology_test.go [hunk 5]
```diff
@@ -278,6 +312,32 @@ func TestFailureDomain(t *testing.T) {
 			},
 			expectedFailureDomain: "rack",
 		},
+		{
+			label: "Case 6", // storagecluster with three or more zone topology labels
```
Comment (jarrpa) [inline pkg/controller/storagecluster/topology_test.go:316]: Isn't this the same as case 4?

**Review by jarrpa — CHANGES_REQUESTED** [review]

**Review by jarrpa — APPROVED** [review]
